File: editor-ace/src/main/java/com/raredev/vcspace/editor/ace/VCSpaceWebInterface.java
package com.raredev.vcspace.editor.ace;

import android.webkit.JavascriptInterface;
import android.widget.Toast;
import com.raredev.vcspace.editor.AceCodeEditor;
import com.raredev.vcspace.events.OnContentChangeEvent;
import org.greenrobot.eventbus.EventBus;

public class VCSpaceWebInterface {
  private AceCodeEditor editor;

  public boolean hasUndo;
  public boolean hasRedo;
  public String value;

  /** Instantiate the interface and set the context. */
  public VCSpaceWebInterface(AceCodeEditor c) {
    editor = c;
  }

  /** Show a toast from the web page. */
  @JavascriptInterface
  public void showToast(String toast) {
    Toast.makeText(editor.getContext(), toast, Toast.LENGTH_SHORT).show();
  }

  @JavascriptInterface
  public void onContentChange(String initialValue, String modifiedValue) {
    editor.setModified(true);
    EventBus.getDefault().post(new OnContentChangeEvent(editor.getFile()));
  }

  @JavascriptInterface
  public void setHasUndo(boolean hasUndo) {
    this.hasUndo = hasUndo;
  }

  @JavascriptInterface
  public void setHasRedo(boolean hasRedo) {
    this.hasRedo = hasRedo;
  }

  @JavascriptInterface
  public void setValue(String value) {
    this.value = value;
  }
}

File: editor-ace/src/main/java/com/raredev/vcspace/editor/ace/VCSpaceWebInterface.kt
package com.raredev.vcspace.editor.ace

import android.webkit.JavascriptInterface
import android.widget.Toast
import com.raredev.vcspace.editor.AceCodeEditor
import com.raredev.vcspace.events.OnContentChangeEvent
import org.greenrobot.eventbus.EventBus

class VCSpaceWebInterface(private val editor: AceCodeEditor) {
  @set:JavascriptInterface var hasUndo = false
  @set:JavascriptInterface var hasRedo = false
  @set:JavascriptInterface var value = ""
  var range: Range? = null
  var cursorPosition: Point? = null

  @JavascriptInterface
  fun showToast(toast: String) = Toast.makeText(editor.context, toast, Toast.LENGTH_SHORT).show()

  @Suppress("UNUSED_PARAMETER")
  @JavascriptInterface
  fun onContentChange(initialValue: String, modifiedValue: String) {
    editor.isModified = true
    EventBus.getDefault().post(OnContentChangeEvent(editor.file))
  }

  @JavascriptInterface
  fun setRange(startRow: Int, startColumn: Int, endRow: Int, endColumn: Int) {
    range = Range(Point(startRow, startColumn), Point(endRow, endColumn))
  }

  @JavascriptInterface
  fun setCursorPosition(row: Int, column: Int) {
    this.cursorPosition = Point(row, column)
  }
}
